_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,606 @@
+}
+
+function lt(UncheckedCounter x, UncheckedCounter y) pure returns (bool) {
+    return UncheckedCounter.unwrap(x) < UncheckedCounter.unwrap(y);
+}
+
+contract C {
+    uint internalCounter = 12;
+
+    function testCounter() public returns (uint) {
+        for (
+            UncheckedCounter i = UncheckedCounter.wrap(12);
+            i < UncheckedCounter.wrap(20);
+            i = i + ONE
+        )
+            ++internalCounter;
+
+        return internalCounter;
+    }
+}
+```
+
+##### Token Conversion
+This is a more complex example that shows multiple aspects of user-defined operators.
+The `TokenConversion` contract holds balances of two different tokens, defined as UDVTs,
+and uses a third type as a common denominator for calculations.
+
+`TokenConversion.sol`
+```solidity
+pragma solidity ^0.8.19;
+
+import {TokenA, TokenB, BaseToken, ZERO, ONE} from "./Tokens.sol";
+
+contract TokenConversion {
+    mapping(address account => TokenA balance) public balanceOfTokenA;
+    mapping(address account => TokenB balance) public balanceOfTokenB;
+
+    BaseToken averageDepositSize = ZERO;
+    BaseToken numberOfDeposits = ZERO;
+
+    function deposit(address receiver, TokenA amount) public {
+        balanceOfTokenA[receiver] = balanceOfTokenA[receiver] + amount;
+        BaseToken depositSize = amount.toBase();
+        averageDepositSize = (numberOfDeposits * averageDepositSize + depositSize) / (numberOfDeposits + ONE);
+        numberOfDeposits = numberOfDeposits + ONE;
+    }
+
+    function averageDepositSizeAsTokenA() public view returns (TokenA) {
+        return averageDepositSize.toA();
+    }
+
+    function averageDepositSizeAsTokenB() public view returns (TokenB) {
+        return averageDepositSize.toB();
+    }
+
+    function convertAToB(TokenA amount) public {
+        address owner = msg.sender;
+        require(balanceOfTokenA[owner] >= amount);
+
+        balanceOfTokenA[owner] = balanceOfTokenA[owner] + -amount;
+        balanceOfTokenB[owner] = balanceOfTokenB[owner] + amount.toB();
+    }
+}
+```
+
+`Tokens.sol`
+```solidity
+pragma solidity ^0.8.19;
+
+import "./TokenOperators.sol";
+
+type TokenA is int;
+using {addA as +, subA as -, unsubA as -, A.toB, A.toBase} for TokenA global;
+using {ltA as <, gteA as >=} for TokenA global;
+
+type TokenB is int;
+using {addB as +} for TokenB global;
+
+type BaseToken is int;
+using {Base.toA, Base.toB} for BaseToken global;
+using {addBase as +, subBase as -, ltBase as <, mulBase as *, divBase as /} for BaseToken global;
+
+BaseToken constant ZERO = BaseToken.wrap(0);
+BaseToken constant ONE = BaseToken.wrap(1);
+```
+
+`TokenOperators.sol`
+```solidity
+pragma solidity ^0.8.19;
+
+import {TokenA, TokenB, BaseToken, ZERO} from "./Tokens.sol";
+
+uint constant TOKEN_A_DIGITS = 10;
+uint constant TOKEN_B_DIGITS = 12;
+
+library A {
+    function toBase(TokenA x) internal pure returns (BaseToken) {
+        return BaseToken.wrap(TokenA.unwrap(x) / int(10 ** TOKEN_A_DIGITS));
```

Comment:
@matheusaaguiar
This conversion loses all fractional digits. It reduces the number to the integer part. I think it would be better to instead scale it up, to something like `max(TOKEN_A_DIGITS, TOKEN_B_DIGITS)`.

This does not affect the operators, but still looks weird.

Reply:
Yeah, I was lacking a better idea and just wrote this... I am going to implement your suggestion.

Reply:
I changed it so that when converting to `BaseToken` the value is upscaled and the reverse happens when converting back to either `A` or `B`.